src/edge/models/fragment_annotator.py

```diff
@@ -15,36 +15,65 @@ def _add_feature(self, name, type, length, strand, qualifiers=None, operation=No
         f.save()
         return f
 
-    def annotate(self, first_base1, last_base1, name, type, strand,
-                 qualifiers=None, operation=None):
-        if self.circular and last_base1 < first_base1:
-            # has to figure out the total length from last chunk
-            length = self.length - first_base1 + 1 + last_base1
-        else:
-            length = last_base1 - first_base1 + 1
-            if length <= 0:
-                raise Exception('Annotation must have length one or more')
+    def annotate_chunk(self, feature, feature_base_first, first_base1, last_base1):
 
         prev_chunk, annotation_start = self._find_and_split_before(first_base1)
         annotation_end, next_chunk = self._find_and_split_before(last_base1 + 1)
 
         # did two splits, so must reload annotation_start in case that got split
         annotation_start = annotation_start.reload()
 
-        new_feature = self._add_feature(name, type, length, strand, qualifiers, operation)
-
         # now, starting with chunk annotation_start, walk through chunks until
         # we hit annotation_end, and add annotation for each chunk
         chunk = annotation_start
-        a_i = 1
+        a_i = 0
         while True:
             fc = self.fragment_chunk(chunk)
-            self._annotate_chunk(chunk, new_feature, a_i, a_i + len(chunk.sequence) - 1)
+            self._annotate_chunk(
+                chunk,
+                feature,
+                feature_base_first + a_i,
+                feature_base_first + a_i + len(chunk.sequence) - 1
+            )
             a_i += len(chunk.sequence)
             if chunk.id == annotation_end.id:
                 break
             chunk = fc.next_chunk
             if chunk is None:
                 chunk = self.start_chunk
 
+    def bp_covered_length(self, first_base1, last_base1):
+        if self.circular and last_base1 < first_base1:
+            # has to figure out the total length from last chunk
+            length = self.length - first_base1 + 1 + last_base1
+        else:
+            length = last_base1 - first_base1 + 1
+            if length <= 0:
+                raise Exception('Annotation must have length one or more')
+        return length
+
+    def annotate_chunks(self, bases, name, type, strand, qualifiers=None, operation=None):
+        bases = sorted(bases, key=lambda t: t[0])
```

Comment:
Wouldn't this mess up circlular annotations? Should this only apply when it's non-circular? And if so, wouldn't it be better to raise an error if someone is giving coordinates where the start is > the end for non-circular sequences?

Reply:
okay, so i don't think it will break circular annotations. however, i do think we should honor the user's ordered list of bps.  i will add a test to show that effect.

Reply:
but if we are annotating, `(987, 123)`, this will reverse the base order list -- that will then mess up the code below that gets the coverage length because it received `(123-987)` instead.